app/main.py

```diff
@@ -86,129 +86,35 @@
-
-    # apply potion if exist
-    if arthur["potion"] is not None:
-        if "power" in arthur["potion"]["effect"]:
-            arthur["power"] += arthur["potion"]["effect"]["power"]
-
-        if "protection" in arthur["potion"]["effect"]:
-            arthur["protection"] += arthur["potion"]["effect"]["protection"]
-
-        if "hp" in arthur["potion"]["effect"]:
-            arthur["hp"] += arthur["potion"]["effect"]["hp"]
-
-    # mordred
-    mordred = knightsConfig["mordred"]
-
-    # apply armour
-    mordred["protection"] = 0
-    for a in mordred["armour"]:
-        mordred["protection"] += a["protection"]
-
-    # apply weapon
-    mordred["power"] += mordred["weapon"]["power"]
-
-    # apply potion if exist
-    if mordred["potion"] is not None:
-        if "power" in mordred["potion"]["effect"]:
-            mordred["power"] += mordred["potion"]["effect"]["power"]
-
-        if "protection" in mordred["potion"]["effect"]:
-            mordred["protection"] += mordred["potion"]["effect"]["protection"]
-
-        if "hp" in mordred["potion"]["effect"]:
-            mordred["hp"] += mordred["potion"]["effect"]["hp"]
-
-    # red_knight
-    red_knight = knightsConfig["red_knight"]
-
-    # apply armour
-    red_knight["protection"] = 0
-    for a in red_knight["armour"]:
-        red_knight["protection"] += a["protection"]
-
-    # apply weapon
-    red_knight["power"] += red_knight["weapon"]["power"]
-
-    # apply potion if exist
-    if red_knight["potion"] is not None:
-        if "power" in red_knight["potion"]["effect"]:
-            red_knight["power"] += red_knight["potion"]["effect"]["power"]
-
-        if "protection" in red_knight["potion"]["effect"]:
-            red_knight["protection"] += red_knight["potion"]["effect"]["protection"]
-
-        if "hp" in red_knight["potion"]["effect"]:
-            red_knight["hp"] += red_knight["potion"]["effect"]["hp"]
-
-    # -------------------------------------------------------------------------------
-    # BATTLE:
-
-    # 1 Lancelot vs Mordred:
+def battle(knights_config: dict) -> dict:
+    list_power = ["power", "protection", "hp"]
+    lancelot = knights_config["lancelot"]
+    mordred = knights_config["mordred"]
+    red_knight = knights_config["red_knight"]
+    arthur = knights_config["arthur"]
+    list_knights = [lancelot, mordred, red_knight, arthur]
+    for knight in list_knights:
+        knight["protection"] = 0
+        for armour_knight in knight["armour"]:
+            knight["protection"] += armour_knight["protection"]
+    for knight_power in list_knights:
+        knight_power["power"] += knight_power["weapon"]["power"]
+    for knight_magic in list_knights:
+        if knight_magic["potion"] is not None:
+            for value in list_power:
+                if value in knight_magic["potion"]["effect"]:
+                    knight_magic[value] += \
+                        knight_magic["potion"]["effect"][value]
```

Comment:
Probably it should be a separate method/function to provide this logic

Reply:
why